### PR TITLE
Cloud spanner

### LINKED ISSUE #161: Implement a Cloud Spanner datastore
Do you have any interest in a Google Spanner integration?

### PR DESCRIPTION
**Pull request by jakedt**
Fixes #161

### CODE REVIEW

File: internal/datastore/common/changes_test.go [hunk 1]
```diff
@@ -311,3 +311,7 @@ func canonicalize(in []*datastore.RevisionChanges) []*datastore.RevisionChanges
 
 	return out
 }
+
+func revisionFromTransactionID(txID uint64) datastore.Revision {
```
Comment (josephschorr) [inline internal/datastore/common/changes_test.go:315]: Move to the top near the other revision key stuff?

File: internal/datastore/common/revisions.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	"context"
+	"time"
+
+	"github.com/rs/zerolog/log"
+	"github.com/shopspring/decimal"
+
+	"github.com/authzed/spicedb/internal/datastore"
+)
+
+type RemoteNowFunction func(context.Context) (datastore.Revision, error)
```
Comment (josephschorr) [inline internal/datastore/common/revisions.go:13]: doc comments on exported items

File: internal/datastore/common/revisions.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+	log.Ctx(ctx).Debug().Time("now", localNow).Time("valid", rcr.revisionValidThrough).Msg("computing new revision")
+
+	nowHLC, err := rcr.NowFunc(ctx)
+	if err != nil {
+		return datastore.NoRevision, err
+	}
+
+	// Round the revision down to the nearest quantization
+	// Apply a delay to enable follower reads: https://www.cockroachlabs.com/docs/stable/follower-reads.html
```
Comment (josephschorr) [inline internal/datastore/common/revisions.go:44]: Does this comment still apply in the common code?

File: internal/datastore/common/sql.go [hunk 4]
```diff
@@ -124,24 +112,22 @@ func (sqf SchemaQueryFilterer) FilterToSubjectFilter(filter *v1.SubjectFilter) S
...
 
 // FilterToUsersets returns a new SchemaQueryFilterer that is limited to resources with subjects
-// in the specified list of usersets.
-func (sqf SchemaQueryFilterer) FilterToUsersets(usersets []*v0.ObjectAndRelation) SchemaQueryFilterer {
+// in the specified list of usersets. Nil or empty usersets parameter does not affect the underlying
+// query.
+func (sqf SchemaQueryFilterer) filterToUsersets(usersets []*v0.ObjectAndRelation) SchemaQueryFilterer {
 	if len(usersets) == 0 {
-		panic("Got empty usersets filter")
+		return sqf
```
Comment (josephschorr) [inline internal/datastore/common/sql.go:130]: Why this change?

File: internal/datastore/spanner/migrations/driver.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	tableSchemaVersion = "schema_version"
+	colVersionNum      = "version_num"
+)
+
+type SpannerMigrationDriver struct {
+	client      *spanner.Client
+	adminClient *admin.DatabaseAdminClient
+}
+
+func NewSpannerDriver(database, credentialsFilePath string) (SpannerMigrationDriver, error) {
```
Comment (josephschorr) [inline internal/datastore/spanner/migrations/driver.go:25]: doc comments on exported items

File: internal/datastore/common/revisions.go (hunk 3 above)
Comment (ecordell) [reply internal/datastore/common/revisions.go:44]: The delay is still there and still configurable, even if it's only useful for CRDB. Maybe we'll find other uses for it in the future?

File: internal/datastore/common/sql.go (hunk 4 above)
Comment (ecordell) [reply internal/datastore/common/sql.go:130]: not sure - looking at where it's used, we shouldn't ever be sending a batch of size 0. I'll play with it a bit.

File: internal/datastore/spanner/gc.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	if sd.config.gcInterval <= 0 {
+		log.Info().Stringer("interval", sd.config.gcInterval).Msg("garbage collection: disabled")
+		return noopCancelFunc
+	}
+
+	log.Info().Stringer("interval", sd.config.gcInterval).Msg("garbage collection: starting")
+
+	s := gocron.NewScheduler(time.UTC)
+	var numRemoved int64
+	_, err := s.Every(sd.config.gcInterval).Do(func() {
```
Comment (ecordell) [inline internal/datastore/spanner/gc.go:28]: We might want some jitter here, or potentially even have the instances coordinate via spanner

File: internal/datastore/spanner/gc.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+		})
+		if err != nil {
+			log.Error().Err(err).Msg("garbage collection: error deleting entries")
+		}
+
+		log.Info().Int64("removed", numRemoved).Stringer("before", oldestRevision).
+			Msg("garbage collection: removed changelog entries")
+	})
+	if err != nil {
+		panic(fmt.Errorf("unable to start garbage collection: %w", err))
```
Comment (ecordell) [inline internal/datastore/spanner/gc.go:59]: instead of panicing, this should thread this back up as an error so that the other goroutines can be cancelled / have a chance to gracefully stop

File: internal/datastore/spanner/gc.go (hunk 6 above)
Comment (josephschorr) [reply internal/datastore/spanner/gc.go:28]: Add a random jitter for now and a comment to make it coordinate?

File: internal/datastore/spanner/gc.go (hunk 7 above)
Comment (josephschorr) [reply internal/datastore/spanner/gc.go:59]: +1

Comment (ecordell) [reply internal/datastore/spanner/gc.go]: Output after doing this:
```
9:46AM INF set log level new level=debug
9:46AM INF set tracing provider new provider=none
9:46AM INF using spanner datastore engine
9:46AM INF gc configured with jitter factor=0.2
9:46AM INF garbage collection: starting interval=3m24.678454111s
9:46AM INF request hedging enabled hedgingQuantile=0.95 initialSlowRequest=10ms maxRequests=1000000
9:46AM DBG configured combined dispatcher upstream=
9:46AM WRN grpc server serving plaintext prefix=grpc
9:46AM INF grpc server started listening addr=:50051 network=tcp prefix=grpc
9:46AM WRN http server serving plaintext addr=:9090 prefix=metrics
9:46AM WRN http server serving plaintext addr=:8080 prefix=dashboard
^C9:46AM INF received interrupt
9:46AM INF shutting down
9:46AM INF http server stopped serving addr=:8080 prefix=dashboard
9:46AM INF http server stopped serving addr=:9090 prefix=metrics
9:46AM INF grpc server stopped listening addr=:50051 network=tcp prefix=grpc
9:46AM INF garbage collection: stopping
```

File: internal/datastore/spanner/gc.go (hunk 6 above)
Comment (ecordell) [reply internal/datastore/spanner/gc.go:26]: Added jitter, I'll make a GH issue for coordination instead of a comment

File: internal/datastore/common/revisions.go (hunk 3 above)
Comment (josephschorr) [reply internal/datastore/common/revisions.go:52]: We might want to add text saying the CRDB is an example

File: internal/datastore/spanner/tuple.go [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+	"github.com/rs/zerolog/log"
+	"google.golang.org/api/iterator"
+
+	"github.com/authzed/spicedb/internal/datastore"
+)
+
+func (sd spannerDatastore) WriteTuples(ctx context.Context, preconditions []*v1.Precondition, mutations []*v1.RelationshipUpdate) (datastore.Revision, error) {
+	ctx, span := tracer.Start(ctx, "WriteTuples")
+	defer span.End()
+	ts, err := sd.client.ReadWriteTransaction(ctx, func(ctx context.Context, rwt *spanner.ReadWriteTransaction) error {
```
Comment (josephschorr) [inline internal/datastore/spanner/tuple.go:22]: newline after the defer

**Review by josephschorr — APPROVED** [review]
LGTM
